### PR TITLE
Widget property generics

### LINKED ISSUE #313: Allow individual widget properties to be extended
Allow individual widget properties to be extended when extending the widget, so instead of:
   `export default class Button extends ButtonBase<ButtonProperties>`
   we would have:
   `export default class Button<P extends ButtonProperties = ButtonProperties> extends ButtonBase<P>`

### LINKED ISSUE #314: Extendable widget properties
**Enhancement** 

Allow each widget's properties interface to be extended, so instead of:
`export default class Button extends ButtonBase<ButtonProperties>`
we would have:
`export default class Button<P extends ButtonProperties = ButtonProperties> extends ButtonBase<P>`

Comes from the first point in https://github.com/dojo/widgets/issues/266#issuecomment-335559039

### PR DESCRIPTION
**Pull request by smhigley**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [ ] Unit or Functional tests are included in the PR

**Description:**

Resolves #314. This is currently blocked by https://github.com/Microsoft/TypeScript/issues/17005, which causes TS errors when narrowing optional properties.

### CODE REVIEW

**Comment by kitsonk** [comment]
Microsoft/TypeScript#17005 will be in TypeScript 2.6 final.

File: src/calendar/example/index.ts [hunk 1]
```diff
@@ -6,6 +6,24 @@ import Calendar from '../../calendar/Calendar';
 import Checkbox from '../../checkbox/Checkbox';
 import dojoTheme from '../../themes/dojo/theme';
 
+interface MyWidgetProperties {
```
Comment (bitpshr) [inline src/calendar/example/index.ts]: Was this intentional, or just for testing?

Comment (smhigley) [reply src/calendar/example/index.ts]: Definitely a mistake :joy:. Thanks for catching that!

**Comment by bitpshr** [comment]
+1 for strictly typing children. I believe most of our applicable components do, like TabController and AccordionPane. Definitely something to stay vigilant about.

**Pull request merged**